Source/NHibernate.Extensions/DeepClone/DeepCloneSessionExtension.cs

```diff
@@ -354,12 +354,31 @@ private static void AddItemToCollection(object newCollection, object collection,
             var enumerable = collection as IEnumerable;
             if (enumerable != null)
             {
-                foreach (var item in enumerable)
+                var enumerableType = enumerable.GetType();
+                if (enumerableType.IsAssignableToGenericType(typeof(IDictionary<,>)))
                 {
-                    addMethod.Invoke(newCollection,
-                                     editBeforeAdding != null
-                                     ? new[] { editBeforeAdding(item) }
-                                     : new[] { item });
+                    var keysProperty = enumerableType.GetProperty("Keys");
+                    var itemProperty = enumerableType.GetProperty("Item");
+                    var keys = keysProperty.GetValue(enumerable) as IEnumerable;
+                    foreach (var key in keys)
+                    {
+                        var value = itemProperty.GetValue(enumerable, new object[] { key });
+                        addMethod.Invoke(newCollection,
+                                         editBeforeAdding != null
+                                         ? new[] { key, editBeforeAdding(value) }
+                                         : new[] { key, value }
+                        );
+                    }
+                }
```

Comment:
I would use `dynamic` rather than reflection as it is faster and would simply the code:
```
if (enumerableType.IsAssignableToGenericType(typeof(IDictionary<,>)))
{
    foreach (dynamic pair in enumerable)
    {
        addMethod.Invoke(newCollection,
            editBeforeAdding != null
                ? new[] {pair.Key, editBeforeAdding(pair.Value)}
                : new[] {pair.Key, pair.Value}
        );
    }
}
```

Reply:
Ok, I'll fix.